### PR TITLE
Expose authorization API

### PR DESCRIPTION
**Pull request by pramodgavali**
https://github.com/IBM/ibmcloud-storage-volume-lib/issues/21

### CODE REVIEW

File: volume-providers/softlayer/backend/network_subnet_ipaddress_service.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	var ntError error
+	ntError = retry(func() error {
+		subnetIP, ntError = ns.networkSubnetIpAddressService.GetObject()
+		return ntError
+	})
+	return subnetIP, ntError
+}
+
+// GetByIPAddress returns Network_Subnet_IpAddress from IP address string
+func (ns *NetworkSubnetIpAddressServiceSL) GetByIPAddress(ipAddress *string) (datatypes.Network_Subnet_IpAddress, error) {
```
Comment (arahamad) [inline volume-providers/softlayer/backend/network_subnet_ipaddress_service.go]: Please add the subject in the function name like `GetSubnetIPFromIPAddress` or `GetSubnetIPByIPAddress`

File: volume-providers/softlayer/common/session.go [hunk 2]
```diff
@@ -330,3 +330,37 @@ func (sls *SLSession) ListAllSnapshots(volumeID string) ([]*provider.Snapshot, e
...
+		err = sls.updateHostIPAuthorization(volumeID, authorizationRequest.HostIps)
+	}
+	sls.Logger.Info("Exit UpdateAuthorization ", zap.Error(err))
+	return err
+
+}
+
+func (sls *SLSession) updateSubnetAuthorization(volumeID int, subnetIDs []string) error {
+	sls.Logger.Info("Entry updateSubnetAuthorization", zap.Reflect("subnetIDs", subnetIDs))
+	subnetList, _ := utils.GetSubnetListFromIDs(sls.Logger, sls.Backend, subnetIDs)
```
Comment (arahamad) [inline volume-providers/softlayer/common/session.go]: Can we check the error also if `GetSubnetListFromIDs` returns error then lets not call other methods

File: volume-providers/softlayer/common/session.go [hunk 3]
```diff
@@ -330,3 +330,37 @@ func (sls *SLSession) ListAllSnapshots(volumeID string) ([]*provider.Snapshot, e
...
+	storageService := sls.Backend.GetNetworkStorageService().ID(volumeID)
+	result, err := storageService.AllowAccessFromSubnetList(subnetList)
+	sls.Logger.Info("Exit updateSubnetAuthorization ", zap.Bool("result", result), zap.Error(err))
+	return err
+}
+
+func (sls *SLSession) updateHostIPAuthorization(volumeID int, hostIPList []string) error {
+	sls.Logger.Info("Entry updateHostIPAuthorization", zap.Reflect("hostIPList", hostIPList))
+	subnetIPAddressList, _ := utils.GetSubnetIpAddressListFromIPs(sls.Logger, sls.Backend, hostIPList)
```
Comment (arahamad) [inline volume-providers/softlayer/common/session.go]: Can we check the error also if `GetSubnetIpAddressListFromIPs ` returns error then lets not call other methods

File: volume-providers/softlayer/utils/storage_utils.go [hunk 4]
```diff
@@ -542,3 +542,35 @@ func ConvertToLocalSnapshotObject(storageSnapshot datatypes.Network_Storage, log
...
+	subnetList := []datatypes.Network_Subnet{}
+	subnetService := session.GetNetworkSubnetService()
+	for _, subnetID := range subnetIDs {
+		subnetIDInt, _ := strconv.Atoi(subnetID)
+		subnet, err := subnetService.ID(subnetIDInt).GetObject()
+		if err == nil {
+			subnetList = append(subnetList, subnet)
+		} else {
+			logger.Error("Unable to get the subnet by id ", zap.Error(err), zap.String("subnetId", subnetID))
+			return nil, err
```
Comment (arahamad) [inline volume-providers/softlayer/utils/storage_utils.go]: Do you mean to say if we are fail to get subnet for any subnetID then we should return nil i.e fail to get for all,  please check in storage plugin what we are doing

File: volume-providers/softlayer/utils/storage_utils.go [hunk 5]
```diff
@@ -542,3 +542,35 @@ func ConvertToLocalSnapshotObject(storageSnapshot datatypes.Network_Storage, log
...
+	return subnetList, nil
+}
+
+//GetSubnetIpAddressListFromIPs returns list of Network_Subnet_IpAddress from IP address list
+func GetSubnetIpAddressListFromIPs(logger *zap.Logger, session backend.Session, ipAddressList []string) ([]datatypes.Network_Subnet_IpAddress, error) {
+	mask := `id,ipAddress,hardware,virtualGuest`
+	subnetIpaddressList := []datatypes.Network_Subnet_IpAddress{}
+	for _, ipAddress := range ipAddressList {
+		subnetIpaddress, err := session.GetNetworkSubnetIpAddressService().Mask(mask).GetByIPAddress(&ipAddress)
+		if err != nil {
```
Comment (arahamad) [inline volume-providers/softlayer/utils/storage_utils.go]: same here, please check how we are doing in are plugin

File: lib/provider/data_types.go [hunk 6]
```diff
@@ -105,3 +105,14 @@ type Snapshot struct {
 	// tags for the snapshot
 	SnapshotTags SnapshotTags `json:"tags"`
 }
+
+// AuthorizationRequest capture details of autorization to be made
+type AuthorizationRequest struct {
+	// Volume to update the authorization
+	Volume Volume
```
Comment (arahamad) [inline lib/provider/data_types.go]: Can we have only ID instead of Volume struct as only volume ID is required.

File: lib/provider/data_types.go [hunk 7]
```diff
@@ -105,3 +105,14 @@ type Snapshot struct {
 	// tags for the snapshot
 	SnapshotTags SnapshotTags `json:"tags"`
 }
+
+// AuthorizationRequest capture details of autorization to be made
+type AuthorizationRequest struct {
```
Comment (arahamad) [inline lib/provider/data_types.go]: Can we have different struct name may be `VolumeAuthRequest`, as per name purpose of this struct is not coming out.

File: lib/provider/data_types.go [hunk 8]
```diff
@@ -105,3 +105,14 @@ type Snapshot struct {
...
+	// Volume to update the authorization
+	Volume Volume
+	// List of subnets to authorize. It might be SubnetIDs or CIDR based on the providers implementaions
+	// For example, IBM Softlyaer provider  expects SubnetIDs to be passed
+	Subnets []string
+	// List of HostIPs to authorize
+	HostIps []string
```
Comment (arahamad) [inline lib/provider/data_types.go]: please change `HostIps` to `HostIPs` similar to the comment

File: lib/provider/volume_manager.go [hunk 9]
```diff
@@ -38,4 +38,8 @@ type VolumeManager interface {
 	// Request Id is the one that is returned when volume is provsioning request is
 	// placed with Iaas provider.
 	GetVolumeByRequestID(requestID string) (*Volume, error)
+
+	//UpdateAuthorization update the authorzation based on authorization RequestTimeout
+	// It removes existing authorization and apply authorization given in the RequestTimeout
+	UpdateAuthorization(authorizationRequest AuthorizationRequest) error
```
Comment (arahamad) [inline lib/provider/volume_manager.go]: Can we have different interface name may be `AuthorizeVolume`, as per name purpose of this method is not coming out.

File: lib/provider/data_types.go (hunk 8 above)
Comment (arahamad) [reply lib/provider/data_types.go]: also can we have json tag as well for all struct filed similar to other struct

File: volume-providers/softlayer/backend/network_subnet_ipaddress_service.go (hunk 1 above)
Comment (pramodgavali) [reply volume-providers/softlayer/backend/network_subnet_ipaddress_service.go]: Need to follow other  functions names -- >Filter,ID,Mask etc.  Also, ideal convention is that if Object name is part of `Struct name` then  its functions name won't have object name again.  e.g `Volume.GetByID() ` (not a `Volume.GetVolumeByID`) You can see Softlayer structs for reference

File: volume-providers/softlayer/common/session.go (hunk 3 above)
Comment (pramodgavali) [reply volume-providers/softlayer/common/session.go]: done

File: volume-providers/softlayer/utils/storage_utils.go (hunk 4 above)
Comment (pramodgavali) [reply volume-providers/softlayer/utils/storage_utils.go]: done

Comment (pramodgavali) [reply volume-providers/softlayer/utils/storage_utils.go]: All-or-None is the common strategy we should follow for any transaction . In can of partial  processing, user will have to keep track on what is failed in what is applied

File: volume-providers/softlayer/utils/storage_utils.go (hunk 5 above)
Comment (pramodgavali) [reply volume-providers/softlayer/utils/storage_utils.go]: All-or-None is the common strategy we should follow for any transaction . In can of partial processing, user will have to keep track on what is failed in what is applied

File: lib/provider/data_types.go (hunk 7 above)
Comment (pramodgavali) [reply lib/provider/data_types.go]: Done --> Changed to `VolumeAuthorization`

File: lib/provider/data_types.go (hunk 8 above)
Comment (pramodgavali) [reply lib/provider/data_types.go]: done

File: lib/provider/volume_manager.go (hunk 9 above)
Comment (pramodgavali) [reply lib/provider/volume_manager.go]: done

File: lib/provider/data_types.go (hunk 6 above)
Comment (pramodgavali) [reply lib/provider/data_types.go]: We want to be as much generic as  possible in defining the interfaces , Not sure which providers wants which all parameters.

**Pull request merged**